presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarArray.java

```diff
@@ -32,6 +32,13 @@ public static ColumnarArray toColumnarArray(Block block)
         if (block instanceof RunLengthEncodedBlock) {
             return toColumnarArray((RunLengthEncodedBlock) block);
         }
+        if (block instanceof LazyBlock) {
+            LazyBlock lazyBlock = (LazyBlock) block;
+            if (!lazyBlock.isLoaded()) {
+                throw new IllegalArgumentException("LazyBlock is expected to be loaded");
```

Comment:
Is this check needed? Isn't that LazyBlock's problem if it cannot load on-demand?

(same in other places)

Reply:
@findepi : Unloaded LazyBlock should never be passed across operators. Before TableScanOperator returns the LazyBlock, it calls `assureLoaded`

Reply:
> Unloaded LazyBlock should never be passed across operators

This is certainly true for `LazyBlock` created by a scan, since it cannot be loaded once it goes out of the door.
My question was because -- technically, we could pass lazy blocks in other cases, right?

Reply:
The problem is the ORC lazy load methods must be called in order, because once you advance the reader the loaders for the previous section will fail.  What I would like to do is to strip the lazy block wrapper in the scan so we never see them out side of the scan operators... they are just extra overhead

Reply:
> The problem is the ORC lazy load methods must be called in order, because once you advance the reader the loaders for the previous section will fail.

Yes! I know unloaded lazy block cannot leave table scan, because it would be defunct. Is table scan the only source of lazy blocks?

Reply:
Yep.  I think we may want to use it in other places in the future... I have been thinking of using lazy function evaluation to effectively stitch operators together without having to actually stitch the code together.